### ISSUE TITLE
🚀 [Feature]: Log show lowerst time-unit that makes sense [#2258]

### ISSUE THREAD

**Issue by the-hotmann** [gofiber/fiber#2561]
### Feature Description

Refering to #2258

When I check my Fiber logs, it instead of `0s` now always shows: `xms`.
This is better, as it is more precise, but I really would recommend it, if Fiber (just like Gin) would show he smallest unit that makes sense:

- `1.423s`
- `456.ms`
- `18.374µs`
- `432ns`

Fiber, will (in my expirience) never show `µs` & `ns`, even though I am relly sure, it does not always take exactly 1ms ;)

### Additional Context (optional)

The log looks like this for me (IP replaced with `123.123.123.123`):

```
12:29:03 | 200 |     5ms |     123.123.123.123 | PUT     | /api/product/1
12:29:03 | 200 |     2ms |     123.123.123.123 | GET     | /api/products
12:29:13 | 200 |     3ms |     123.123.123.123 | GET     | /admin
12:29:13 | 304 |      0s |     123.123.123.123 | GET     | /app.min.css
12:29:13 | 200 |     1ms |     123.123.123.123 | GET     | /api/categories
12:29:13 | 200 |     2ms |     123.123.123.123 | GET     | /api/products
12:29:23 | 200 |      0s |     123.123.123.123 | GET     | /api/product/1
```

Also the responsetime of the very same request fluctuates between `0s` and `5ms` - which in my case is very high and other WebFrameworks do not do.
Since I migrate to Fiber I hope that these little things will be fixed somewhen, it is not urgent, but also the Benchmarks (where Fiber always outperforms others) are ATM not matching my real-life expirience, of beeing any faster than Gin for example.
Maybe they just apply, when a big number of requests are triggered.

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my suggestion prior to opening this one.
- [X] I understand that improperly formatted feature requests may be closed without explanation.

**Comment by sixcolors** [comment]
gin-gonic/gin defaultLogFormatter https://github.com/gin-gonic/gin/blob/62b50cfbc0de877207ff74c160a23dff6394f563/logger.go#L143 used %13v on params.Latency.

which looks like:

```console
|       4.333µs |
|    1.133916ms |
|  1.000057458s |
```

gofiber/fiber/middleware/logger uses %7v on TagLatency.latency

```console
| 4.834µs |
| 1.136834ms |
| 1.000066333s |
```

13 chars seems to be better.

Otherwise gofiber/fiber/middleware/logger behaves the same as the default Gin logger.

I will add a test and change the formating of the string to %13v in a PR.

**Comment by sixcolors** [comment]
@MartinHotmann #2569 should resolve padding, but current code does work per gin defaultLogFormatter.

**Comment by MartinHotmann** [comment]
@sixcolors thanks for your response!

For me it was never about the padding, but thanks for that - I appreciate it. For me it was about:

Fiber never shows any "`request process times`" which are lower than `1ms` or `0s`. The lowest it goes for me at sending back a static string is `1ms` or sometimes it shows `0s`.

(you can see this in the example I provided above)

But here again. I will provide a log from fiber when calling a static site:

```
23:04:53 | 200 |     7ms |     172.18.0.30 | GET     | /
23:05:05 | 200 |     2ms |     172.18.0.30 | GET     | /warum
23:05:06 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:07 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:07 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:07 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:07 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:07 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:07 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:08 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:09 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:09 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:09 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:09 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:10 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:10 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:10 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:10 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:10 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:10 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:11 | 200 |      0s |     172.18.0.30 | GET     | /warum
23:05:11 | 200 |      0s |     172.18.0.30 | GET     | /warum
```

Please note, that the page `/warum` is completely static.

The first time I call it, it takes a little longer, but the other times are much faster, but still it shows `0s`.
Yes, Gin formats the time the same, but they actually show me how many `µs` or even `ns` it took to process that request in total.


Since we are already talking:

is it possible to pass that string/variable to the template?
I would like to render that exact time on the website, which then shows how long/short it took.


Thanks in advance :)

**Comment by sixcolors** [comment]
It should be doing what you want. Let me invesitgate further.

**Comment by sixcolors** [comment]
@MartinHotmann I can't seem to replicate your issue with logging the latency.

```console
 ┌───────────────────────────────────────────────────┐ 
 │                   Fiber v2.48.0                   │ 
 │               http://127.0.0.1:8003               │ 
 │       (bound on host 0.0.0.0 and port 8003)       │ 
 │                                                   │ 
 │ Handlers ............. 4  Processes ........... 1 │ 
 │ Prefork ....... Disabled  PID ............. 64375 │ 
 └───────────────────────────────────────────────────┘ 

22:13:06 [127.0.0.1] 200 -    1.218709ms GET /
22:13:08 [127.0.0.1] 200 -    1.282416ms GET /
22:13:08 [127.0.0.1] 200 -    1.254958ms GET /
22:13:09 [127.0.0.1] 200 -    1.389375ms GET /
22:13:09 [127.0.0.1] 200 -     105.792µs GET /
22:13:10 [127.0.0.1] 200 -     144.666µs GET /
22:13:10 [127.0.0.1] 200 -     140.709µs GET /
22:13:11 [127.0.0.1] 200 -     146.792µs GET /
22:13:12 [127.0.0.1] 200 -    1.288667ms GET /
22:13:11 [127.0.0.1] 200 -  1.001153959s GET /
22:13:12 [127.0.0.1] 200 -  1.000757084s GET /
22:13:13 [127.0.0.1] 200 -   1.00119775s GET /
```

here is the sample app I used:

```go
package main

import (
	"math/rand"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/logger"
)

func main() {
	server := fiber.New()
	server.Use(logger.New(logger.Config{
		Format: "${time} [${ip}] ${status} - ${latency} ${method} ${path}\n",
	}))

	server.Static("/", "../static", fiber.Static{
		Compress: true,
	})

	durations := []time.Duration{time.Nanosecond, time.Microsecond, time.Millisecond, time.Second}
	rand.Seed(time.Now().UnixNano())

	server.Get("/", func(c *fiber.Ctx) error {
		// randomly sleep for either 1ns, 1μs, 1ms, or 1s
		duration := durations[rand.Intn(len(durations))]
		time.Sleep(duration)
		return c.SendString("Hello, World!")
	})
	server.Listen(":8003")
}
```

What version of fiber are you using and on what platform?

What is your logger config?

-----------

When dealing with web requests, it's important to recognize that the time taken to render a response in the template is a separate and subsequent event to the request and its processing. As such, you cannot directly use the latency measured by the logger middleware, which encompasses the entire request lifecycle (from start to response), and display it within the same response content. This would indeed create a paradoxical situation.

To display the latency measured by the logger middleware on your website, you would typically need to either calculate the latency again within your route handler (as previously mentioned) or log it to an external monitoring or analytics system.

You could track average latency by doing something like this:

```go
type customWriter struct {
	stdout io.Writer
	avgLat time.Duration
	count  int
	mux    sync.Mutex
}

func (d *customWriter) Write(p []byte) (n int, err error) {
	// write to stdout
	n, err = d.stdout.Write(p)

	if err != nil {
		return n, err
	}

	// this is a hacky way to do it, and you probably want something more sophisticated.
	// Get latency from Format: "${time} [${ip}] ${status} - ${latency} ${method} ${path}\n",
	regexp := regexp.MustCompile(`([0-9]+(\.[0-9]+)?(ns|µs|ms|s))`)
	matches := regexp.FindStringSubmatch(string(p))
	if len(matches) > 0 {
		latency, err := time.ParseDuration(matches[1])
		if err != nil {
			return n, err
		}
		d.mux.Lock()
		if d.count == 0 {
			d.avgLat = latency
		} else {
			d.avgLat = ((d.avgLat * time.Duration(d.count)) + latency) / time.Duration(d.count+1)
		}
		d.count++
		d.mux.Unlock()
	}

	return n, nil
}

func main() {
	server := fiber.New()
	customWriter := &customWriter{
		stdout: os.Stdout,
		avgLat: -1.000,
	}

	server.Use(logger.New(logger.Config{
		Format: "${time} [${ip}] ${status} - ${latency} ${method} ${path}\n",
		Output: customWriter,
	}))

	...

	server.Get("/", func(c *fiber.Ctx) error {
		return c.SendString("Hello, World! The Average Latency is: " + customWriter.avgLat.String())
	})
	server.Listen(":8003")
}
```

However you should probably consider sending latency measurements to an external metrics system, like Prometheus or StatsD, which are designed to handle metrics tracking at scale without impacting the main application's performance.

**Comment by sixcolors** [comment]
@MartinHotmann. If you aren't aware, Fiber offers a built-in middleware called [monitor](https://docs.gofiber.io/api/middleware/monitor/) that might partially serve your purpose. It's worth exploring this middleware as it provides monitoring functionalities that could potentially enhance your ability to track and display latency and other metrics on your website.

**Comment by gaby** [comment]
There's also the `fiberprometheus` middleware, that's used by a lot of projects. It records latencies and includes a grafana dashboard.

https://github.com/ansrivas/fiberprometheus

**Comment by MartinHotmann** [comment]
@sixcolors 

```
server.Use(logger.New(logger.Config{
	Format: "${time} [${ip}] ${status} - ${latency} ${method} ${path}\n",
}))
```

You are right - this did the trick!

I was using:

```
app.Use(logger.New())
```


So it seems like that the default config is not providing the wanted output.
Thanks for pointing that out - I checked the default config and I am fine with it, as it also includes the latency variable - but it renders it differently.

And also thanks for https://github.com/gofiber/fiber/pull/2569 this should get rif, of the not properly formatted logs.

@gaby thanks for the recommendation, but I guess this is a little overkill for my small project. Will look into anyway!

**Comment by sixcolors** [comment]
@the-hotmann It seems I've figured out the issue. When you utilize the default formatter, the code within the tags.go file are not utilized. Instead, the output functions located in logger.go are used. It seems this was done for the purpose of adding colors; the duplication seems somewhat unnecessary and I may refactor the code at a later time. For now I have submitted a pull request to address the issue in PR #2580.
